spam/03_spam_data_slicing_tutorial.py

```diff
@@ -352,93 +239,166 @@ def textblob_polarity(x):
+# Using Snorkel's [`Trainer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.Trainer.html), we fit to `train_dl`, and validate on `valid_dl`.
 #
-# A technique like upsampling might work with a small number of slices, but with hundreds or thousands or production slices, it could quickly become intractable to tune upsampling weights per slice.
-# In the following section, we show a modeling approach that we call _Slice-based Learning,_ which handles numerous slices using with slice-specific representation learning.
+# We note that we can monitor slice-specific performance during training — this is a powerful way to track especially critical subsets of the data.
+# If logging in `Tensorboard` (i.e. [`snorkel.classification.TensorboardWritier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.TensorBoardWriter.html)), we would visualize individual loss curves and validation metrics to debug convegence for specific slices.
+
+# %%
+from snorkel.classification import Trainer
+
+# For demonstration purposes, we set n_epochs=2
+trainer = Trainer(lr=1e-4, n_epochs=2)
+trainer.fit(slice_model, [train_dl, valid_dl])
 
 # %% [markdown]
 # ### Representation learning with slices
 
 # %% [markdown]
 # To cope with scale, we will attempt to learn and combine many slice-specific representations with an attention mechanism.
-# (For details, please see our technical report — coming soon!)
+# (For details about this approach, please see our technical report — coming soon!)
 
 # %% [markdown]
-# First, we update our existing dataloader with our new slices.
+# First, we'll generate the remaining `S` matrixes with the new set of slicing functions.
 
 # %%
 applier = PandasSFApplier(sfs)
 S_train = applier.apply(df_train)
 S_valid = applier.apply(df_valid)
 
-# %%
-slice_names = [sf.name for sf in sfs]
-add_slice_labels(dl_train, spam_task, S_train, slice_names)
-add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
-add_slice_labels(dl_test, spam_task, S_test, slice_names)
-
 # %% [markdown]
-# Using the helper, [`convert_to_slice_tasks`](https://snorkel.readthedocs.io/en/master/packages/_autosummary/slicing/snorkel.slicing.convert_to_slice_tasks.html), we can convert our original `spam_task` into slice-specific tasks.
-# These will be used to learn "expert task heads" for each slice, in the style of multi-task learning.
-# The original `spam_task` now contains the attention mechanism to then combine these slice experts.
+# We now highlight the slice-aware capabilities of `SlicingClassifier`.
+# At a high-level, this model adds additional capacity corresponding to each slice through additional _slice-specific tasks_.
 
 # %%
-from snorkel.slicing import convert_to_slice_tasks
+slice_model = SlicingClassifier(
+    base_architecture=mlp,
+    head_dim=bow_dim,
+    slice_names=slice_names,
+)
```

Comment:
I guess, to emphasize what I was saying before, it seems odd to have two different slice models defined in such close succession?  Why not just define one above, then talk about both the monitoring-during-training and slice-aware capabilities in one go?

This also makes it less of a sore thumb that results don't necessarily improve imo